File: Level_3/징검다리 건너기.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool midCheck(int mid, const vector<int>& stones, int k) {
    int cnt = 0;
    for(int i = 0; i < stones.size(); ++i) {
        if(stones[i] - mid <= 0) {
            cnt++;
            if(cnt >= k) return false;
        }
        else cnt = 0;
    }
    return true;
}

int binsearch(const vector<int>& stones, int k) {
    int lo = 0, hi = *max_element(stones.begin(), stones.end());
    while(lo+1 < hi) {
        int mid = (lo + hi) / 2;
        if(midCheck(mid, stones, k)) 
            lo = mid;
        else
            hi = mid;
    }
    return hi;
}

int solution(vector<int> stones, int k) {
    int answer = binsearch(stones, k);
    return answer;
}

File: 2019 카카오 개발자 겨울 인턴십/징검다리 건너기.java
class Solution {
    boolean check(int[] stones, int mid, int k) {
        int cnt = 0;
        for(int i = 0; i < stones.length; ++i) {
            if(stones[i] - mid <= 0) cnt++;
            else cnt = 0;
            if(cnt == k) return false;
        }
        return true;
    }
    
    public int solution(int[] stones, int k) {
        int answer = 0;
        int lo = 0, hi = 0;
        for(int stone : stones) 
            hi = Math.max(hi, stone);
        while(lo+1 < hi) {
            int mid = (lo + hi) / 2;
            if(check(stones, mid, k))
                lo = mid;
            else
                hi = mid;
        }
        return hi;
    }
}
